### ISSUE TITLE
abs, real, imag aren't type-stable for special matrices/don't exist

### ISSUE THREAD

**Issue by kshyatt** [JuliaLang/julia#12910]
e.g.
```julia
julia> A = Tridiagonal(randn(5),randn(6),randn(5))
6x6 Base.LinAlg.Tridiagonal{Float64}:
 -2.48095    0.645111   0.0         0.0        0.0       0.0
  0.481481  -0.822526  -0.0483536   0.0        0.0       0.0
  0.0       -1.69165    0.0961839   0.069047   0.0       0.0
  0.0        0.0       -0.0336218   0.825325  -2.58841   0.0
  0.0        0.0        0.0        -0.58766    0.778614  0.278351
  0.0        0.0        0.0         0.0        0.446866  1.64492

julia> abs(A)
6x6 Array{Float64,2}:
 2.48095   0.645111  0.0        0.0       0.0       0.0
 0.481481  0.822526  0.0483536  0.0       0.0       0.0
 0.0       1.69165   0.0961839  0.069047  0.0       0.0
 0.0       0.0       0.0336218  0.825325  2.58841   0.0
 0.0       0.0       0.0        0.58766   0.778614  0.278351
 0.0       0.0       0.0        0.0       0.446866  1.64492
```

```julia
julia> A = Diagonal(rand(5) + im*rand(5))
5x5 Base.LinAlg.Diagonal{Complex{Float64}}:
 0.496984+0.316435im       0.0+0.0im             0.0+0.0im             0.0+0.0im          0.0+0.0im
      0.0+0.0im       0.508754+0.135719im        0.0+0.0im             0.0+0.0im          0.0+0.0im
      0.0+0.0im            0.0+0.0im       0.0200631+0.0938123im       0.0+0.0im          0.0+0.0im
      0.0+0.0im            0.0+0.0im             0.0+0.0im        0.091229+0.717726im     0.0+0.0im
      0.0+0.0im            0.0+0.0im             0.0+0.0im             0.0+0.0im       0.9047+0.479762im

julia> real(A)
ERROR: MethodError: `real` has no method matching real(::Base.LinAlg.Diagonal{Complex{Float64}})
```
Since implementing these violates the feature freeze, I didn't open a PR. But after 0.4 is out we might want to think about adding these methods.

**Comment by ViralBShah** [comment]
This seems a bit on the border, but unless someone really wants it, I think I would wait for the branch as well.

**Comment by kshyatt** [comment]
Closed by #13049
